File: hostel_management_system_backend/Services/AuthService.cs
using Isopoh.Cryptography.Argon2;
using hostel_management_system_backend.Exceptions;

public interface IAuthService
{
    Task<AuthTokenIssueResult?> LoginAsync(LoginRequestDto dto, CancellationToken cancellationToken);
    Task<AuthTokenIssueResult?> RefreshAsync(string refreshToken, CancellationToken cancellationToken);
    Task LogoutAsync(string refreshToken, CancellationToken cancellationToken);
}

public sealed class AuthService : IAuthService
{
    private readonly IAuthRepository _repo;
    private readonly JwtService _jwtService;
    private readonly IConfiguration _configuration;

    public AuthService(IAuthRepository repo, JwtService jwtService, IConfiguration configuration)
    {
        _repo = repo;
        _jwtService = jwtService;
        _configuration = configuration;
    }

    public async Task<AuthTokenIssueResult?> LoginAsync(LoginRequestDto dto, CancellationToken cancellationToken)
    {
        var user = await _repo.GetUserByEmailForUpdateAsync(dto.Email, cancellationToken);

        if (user is null)
        {
            throw new UnauthorizedException("Invalid email or password.");
        }

        if (!Argon2.Verify(user.PasswordHash, dto.Password))
        {
            throw new UnauthorizedException("Invalid email or password.");
        }

        user.LastActivityAt = DateTime.UtcNow;

        var accessToken = _jwtService.GenerateAccessToken(user);
        var refreshTokenValue = _jwtService.GenerateRefreshToken();
        var refreshTokenHash = RefreshTokenHasher.Hash(refreshTokenValue);
        var refreshExpiresAt = DateTime.UtcNow.Add(GetRefreshLifetime(user.Role, dto.RememberMe));

        var refreshToken = new RefreshToken
        {
            UserId = user.Id,
            TokenHash = refreshTokenHash,
            ExpiresAt = refreshExpiresAt,
            RememberMe = dto.RememberMe,
            Revoked = false,
            CreatedAt = DateTime.UtcNow
        };

        await _repo.AddRefreshTokenAsync(refreshToken, cancellationToken);
        await _repo.SaveChangesAsync(cancellationToken);

        return new AuthTokenIssueResult(
            new AuthTokensResponseDto(
                accessToken.Token,
                accessToken.ExpiresAt,
                user.Id,
                user.Email,
                user.Role
            ),
            refreshTokenValue,
            refreshExpiresAt
        );
    }

    public async Task<AuthTokenIssueResult?> RefreshAsync(string refreshToken, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(refreshToken))
        {
            throw new UnauthorizedException("Refresh token is missing.");
        }

        var refreshTokenHash = RefreshTokenHasher.Hash(refreshToken);

        var tokenEntity = await _repo.GetRefreshTokenWithUserForUpdateAsync(refreshTokenHash, cancellationToken);

        if (tokenEntity is null || tokenEntity.Revoked || tokenEntity.ExpiresAt <= DateTime.UtcNow)
        {
            throw new UnauthorizedException("Session expired. Please log in again.");
        }

        if (tokenEntity.User.Role == UserRole.Admin)
        {
            var idleTimeoutMinutes = GetIntOrDefault(_configuration["AuthSettings:AdminIdleTimeoutMinutes"], 30);
            var idleTime = DateTime.UtcNow - tokenEntity.User.LastActivityAt;
            if (idleTime > TimeSpan.FromMinutes(idleTimeoutMinutes))
            {
                tokenEntity.Revoked = true;
                await _repo.SaveChangesAsync(cancellationToken);
                throw new UnauthorizedException("Session expired due to inactivity. Please log in again.");
            }
        }

        tokenEntity.Revoked = true;

        var newRefreshTokenValue = _jwtService.GenerateRefreshToken();
        var newRefreshTokenHash = RefreshTokenHasher.Hash(newRefreshTokenValue);
        var newRefreshExpiresAt = DateTime.UtcNow.Add(GetRefreshLifetime(tokenEntity.User.Role, tokenEntity.RememberMe));

        var newTokenEntity = new RefreshToken
        {
            UserId = tokenEntity.UserId,
            TokenHash = newRefreshTokenHash,
            ExpiresAt = newRefreshExpiresAt,
            RememberMe = tokenEntity.RememberMe,
            Revoked = false,
            CreatedAt = DateTime.UtcNow
        };

        var accessToken = _jwtService.GenerateAccessToken(tokenEntity.User);

        await _repo.AddRefreshTokenAsync(newTokenEntity, cancellationToken);
        await _repo.SaveChangesAsync(cancellationToken);

        return new AuthTokenIssueResult(
            new AuthTokensResponseDto(
                accessToken.Token,
                accessToken.ExpiresAt,
                tokenEntity.User.Id,
                tokenEntity.User.Email,
                tokenEntity.User.Role
            ),
            newRefreshTokenValue,
            newRefreshExpiresAt
        );
    }

    public async Task LogoutAsync(string refreshToken, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(refreshToken))
        {
            return;
        }

        var refreshTokenHash = RefreshTokenHasher.Hash(refreshToken);

        var tokenEntity = await _repo.GetActiveRefreshTokenForUpdateAsync(refreshTokenHash, cancellationToken);

        if (tokenEntity is null)
        {
            return;
        }

        tokenEntity.Revoked = true;
        await _repo.SaveChangesAsync(cancellationToken);
    }

    private TimeSpan GetRefreshLifetime(UserRole role, bool rememberMe)
    {
        if (role == UserRole.Admin)
        {
            var adminHours = GetIntOrDefault(_configuration["AuthSettings:AdminRefreshExpiryHours"], 12);
            return TimeSpan.FromHours(adminHours);
        }

        var userDays = rememberMe
            ? GetIntOrDefault(_configuration["AuthSettings:UserRefreshExpiryRememberDays"], 30)
            : GetIntOrDefault(_configuration["AuthSettings:UserRefreshExpiryDays"], 1);

        return TimeSpan.FromDays(userDays);
    }

    private static int GetIntOrDefault(string? value, int fallback)
    {
        return int.TryParse(value, out var parsed) ? parsed : fallback;
    }
}

public sealed record AuthTokenIssueResult(
    AuthTokensResponseDto Response,
    string RefreshToken,
    DateTime RefreshTokenExpiresAt
);

File: src/auth/auth.service.ts
import { createHash, randomBytes } from 'node:crypto';

import { Injectable } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import argon2 from 'argon2';

import {
  AppBadRequestException,
  AppConflictException,
  AppUnauthorizedException,
} from '../common/exceptions/app-exception';
import { tryParseUserRole, UserRole, userRoleToName } from '../common/enums/app.enums';
import { AppConfigService } from '../config/app-config.service';
import { DatabaseService } from '../database/database.service';
import { UserRecord } from '../database/database.schemas';
import { UserRegisterDto } from '../users/dto/users.dto';
import { AuthTokenIssueResult, LoginRequestDto } from './dto/auth.dto';
import { JwtPayload } from './interfaces/jwt-payload.interface';

@Injectable()
export class AuthService {
  constructor(
    private readonly db: DatabaseService,
    private readonly jwtService: JwtService,
    private readonly configService: AppConfigService,
  ) {}

  async login(dto: LoginRequestDto): Promise<AuthTokenIssueResult> {
    const user = await this.db.user.findFirst({
      where: {
        email: dto.email,
        isDeleted: false,
      },
    });

    if (!user) {
      throw new AppUnauthorizedException('Invalid email or password.');
    }

    const passwordValid = await argon2.verify(user.passwordHash, dto.password);
    if (!passwordValid) {
      throw new AppUnauthorizedException('Invalid email or password.');
    }

    await this.db.user.update({
      where: { id: user.id },
      data: { lastActivityAt: new Date() },
    });

    const rememberMe = user.role === UserRole.Admin ? false : Boolean(dto.rememberMe);
    return this.issueTokens(user, rememberMe);
  }

  async refresh(refreshToken: string): Promise<AuthTokenIssueResult> {
    if (!refreshToken?.trim()) {
      throw new AppUnauthorizedException('Refresh token is missing.');
    }

    const tokenHash = this.hashRefreshToken(refreshToken);
    const tokenEntity = await this.db.refreshToken.findFirst({
      where: { tokenHash },
      include: { user: true },
    });

    if (!tokenEntity || tokenEntity.revoked || tokenEntity.expiresAt <= new Date() || tokenEntity.user.isDeleted) {
      throw new AppUnauthorizedException('Session expired. Please log in again.');
    }

    if (tokenEntity.user.role === UserRole.Admin) {
      const idleTimeoutMs = this.configService.adminIdleTimeoutMinutes * 60_000;
      const idleDuration = Date.now() - tokenEntity.user.lastActivityAt.getTime();
      if (idleDuration > idleTimeoutMs) {
        await this.db.refreshToken.update({
          where: { id: tokenEntity.id },
          data: { revoked: true },
        });
        throw new AppUnauthorizedException('Session expired due to inactivity. Please log in again.');
      }
    }

    await this.db.refreshToken.update({
      where: { id: tokenEntity.id },
      data: { revoked: true },
    });

    return this.issueTokens(tokenEntity.user, tokenEntity.rememberMe);
  }

  async logout(refreshToken?: string | null) {
    if (!refreshToken?.trim()) {
      return;
    }

    const tokenHash = this.hashRefreshToken(refreshToken);
    await this.db.refreshToken.updateMany({
      where: {
        tokenHash,
        revoked: false,
      },
      data: {
        revoked: true,
      },
    });
  }

  async register(dto: UserRegisterDto): Promise<AuthTokenIssueResult> {
    const existingUser = await this.db.user.findFirst({
      where: { email: dto.email },
    });

    if (existingUser) {
      throw new AppConflictException('A user with this email already exists.', 'email_conflict');
    }

    let role = UserRole.Student;
    if (dto.role !== undefined && dto.role !== null) {
      const parsedRole = tryParseUserRole(dto.role);
      if (parsedRole === null || parsedRole === UserRole.Admin) {
        throw new AppBadRequestException(
          'Invalid role. Only Student and Owner can self-register.',
          'invalid_role',
        );
      }

      role = parsedRole;
    }

    let user: UserRecord;
    try {
      user = await this.db.user.create({
        data: {
          fullName: dto.fullName,
          email: dto.email,
          phoneNumber: dto.phoneNumber,
          passwordHash: await argon2.hash(dto.password),
          role,
          lastActivityAt: new Date(),
          createdAt: new Date(),
          isDeleted: false,
        },
      });
    } catch {
      throw new AppConflictException('A user with this email already exists.', 'email_conflict');
    }

    return this.issueTokens(user, false);
  }

  private async issueTokens(user: UserRecord, rememberMe: boolean): Promise<AuthTokenIssueResult> {
    const accessToken = await this.generateAccessToken(user);
    const refreshToken = this.generateRefreshToken();
    const refreshTokenHash = this.hashRefreshToken(refreshToken);
    const refreshTokenExpiresAt = new Date(Date.now() + this.getRefreshLifetimeMs(user.role, rememberMe));

    await this.db.refreshToken.create({
      data: {
        userId: user.id,
        tokenHash: refreshTokenHash,
        expiresAt: refreshTokenExpiresAt,
        rememberMe,
        revoked: false,
        createdAt: new Date(),
      },
    });

    return {
      response: {
        accessToken: accessToken.token,
        accessTokenExpiresAt: accessToken.expiresAt,
        userId: user.id,
        fullName: user.fullName,
        email: user.email,
        role: user.role as UserRole,
      },
      refreshToken,
      refreshTokenExpiresAt,
    };
  }

  private async generateAccessToken(user: UserRecord): Promise<{ token: string; expiresAt: Date }> {
    const expiresInMinutes =
      user.role === UserRole.Admin
        ? this.configService.adminAccessExpiryMinutes
        : this.configService.userAccessExpiryMinutes;

    const expiresAt = new Date(Date.now() + expiresInMinutes * 60_000);
    const payload: JwtPayload = {
      sub: user.id,
      email: user.email,
      role: userRoleToName(user.role as UserRole),
    };

    const token = await this.jwtService.signAsync(payload, {
      secret: this.configService.jwtSecret,
      issuer: this.configService.jwtIssuer,
      audience: this.configService.jwtAudience,
      expiresIn: `${expiresInMinutes}m`,
    });

    return { token, expiresAt };
  }

  private generateRefreshToken(): string {
    return randomBytes(64).toString('base64');
  }

  private hashRefreshToken(refreshToken: string): string {
    let tokenBuffer: Buffer;

    try {
      tokenBuffer = Buffer.from(refreshToken, 'base64');
    } catch {
      tokenBuffer = Buffer.from(refreshToken, 'utf8');
    }

    return createHash('sha256').update(tokenBuffer).digest('hex');
  }

  private getRefreshLifetimeMs(role: number, rememberMe: boolean): number {
    if (role === UserRole.Admin) {
      return this.configService.adminRefreshExpiryHours * 60 * 60 * 1000;
    }

    const days = rememberMe ? this.configService.userRefreshRememberDays : this.configService.userRefreshExpiryDays;
    return days * 24 * 60 * 60 * 1000;
  }
}
